h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -199,22 +352,39 @@ public void cv_computeAndSetOptimalParameters(ModelBuilder[] cvModelBuilders) {
           for (Key k : g._toRemove)
             Keyed.remove(k);
       }
-      _parms._lambda = Arrays.copyOf(_parms._lambda, lmin_max + 1);
-      _xval_deviances = Arrays.copyOf(_xval_deviances, lmin_max + 1);
-      _xval_sd = Arrays.copyOf(_xval_sd, lmin_max + 1);
+
       for (int i = 0; i < cvModelBuilders.length; ++i) {
         GLM g = (GLM) cvModelBuilders[i];
-        g._model._output.setSubmodelIdx(bestId);
+        g._model._output.setSubmodelIdx(bestId, g._parms);
       }
+
       double bestDev = _xval_deviances[bestId];
       double bestDev1se = bestDev + _xval_sd[bestId];
-      int bestId1se = bestId;
-      while (bestId1se > 0 && _xval_deviances[bestId1se - 1] <= bestDev1se)
-        --bestId1se;
-      _lambdaCVEstimate = ((GLM) cvModelBuilders[0])._model._output._submodels[bestId].lambda_value;
-      _bestCVSubmodel = bestId;
-      _model._output._lambda_1se = bestId1se; // submodel ide with bestDev+one sigma
-      _model._output._selected_submodel_idx = bestId; // set best submodel id here
+      int finalBestId = bestId;
+      Integer[] orderedLambdaIndices = IntStream
+                .range(0, lmin_max)
+                .filter(i -> alphasAndLambdas[i] == alphasAndLambdas[finalBestId]) // get just the lambdas corresponding to the selected alpha
+                .boxed()
+                .sorted((a,b) -> (int) Math.signum(alphasAndLambdas[b+numOfSubmodels] - alphasAndLambdas[a+numOfSubmodels]))
+                .toArray(Integer[]::new);
+      int bestId1se = IntStream
+                .range(0, orderedLambdaIndices.length)
+                .filter(i -> orderedLambdaIndices[i] == finalBestId)
+                .findFirst()
+                .orElse(orderedLambdaIndices.length - 1);
+
+      while (bestId1se > 0 && _xval_deviances[orderedLambdaIndices[bestId1se - 1]] <= bestDev1se)
+          --bestId1se;
+      // get the index into _parms.lambda/_xval_deviances etc
+      _lambdaCVEstimate = alphasAndLambdas[numOfSubmodels + bestId];
+      bestId1se = orderedLambdaIndices[bestId1se];
+      _model._output._lambda_1se = alphasAndLambdas[numOfSubmodels + bestId1se]; // submodel ide with bestDev+one sigma
+
+      // set the final selected alpha and lambda
+      _parms._alpha = new double[] {alphasAndLambdas[bestId]};
+      _parms._lambda = new double[] {alphasAndLambdas[numOfSubmodels + bestId]};
+      _model._output._selected_submodel_idx = 0; // set best submodel id here
+
```

Comment:
@tomasfryda  If users turned off lambda_search, I will agree this is the correct approach to run the main model with just the best alpha lambda determined by the best average test deviance across all models.  

However, with lambda_search, each sequential model training is based off the last one.  Hence, in this case, instead of running just one lambda value, we need to run from the highest lambda value to the best lambda value determined by the average test deviance and go down one more lambda below (not sure is this is what was implemented before).  You can't just use one lambda value here.